runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/FusedPipeline.java

```diff
@@ -19,54 +19,83 @@
 package org.apache.beam.runners.core.construction.graph;
 
 import com.google.auto.value.AutoValue;
+import com.google.common.collect.Sets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
 import org.apache.beam.model.pipeline.v1.RunnerApi.Components;
-import org.apache.beam.model.pipeline.v1.RunnerApi.Components.Builder;
 import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
 import org.apache.beam.model.pipeline.v1.RunnerApi.Pipeline;
 import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
 
-/**
- * A {@link Pipeline} which has been separated into collections of executable components.
- */
+/** A {@link Pipeline} which has been separated into collections of executable components. */
 @AutoValue
 public abstract class FusedPipeline {
   static FusedPipeline of(
       Set<ExecutableStage> environmentalStages, Set<PTransformNode> runnerStages) {
     return new AutoValue_FusedPipeline(environmentalStages, runnerStages);
   }
 
-  /**
-   * The {@link ExecutableStage executable stages} that are executed by SDK harnesses.
-   */
+  /** The {@link ExecutableStage executable stages} that are executed by SDK harnesses. */
   public abstract Set<ExecutableStage> getFusedStages();
 
-  /**
-   * The {@link PTransform PTransforms} that a runner is responsible for executing.
-   */
+  /** The {@link PTransform PTransforms} that a runner is responsible for executing. */
   public abstract Set<PTransformNode> getRunnerExecutedTransforms();
 
+  public RunnerApi.Pipeline toPipeline(Components initialComponents) {
+    Map<String, PTransform> executableTransforms = getExecutableTransforms(initialComponents);
+    Components fusedComponents = initialComponents.toBuilder()
+        .putAllTransforms(executableTransforms)
+        .putAllTransforms(getFusedTransforms())
+        .build();
+    List<String> rootTransformIds =
+        StreamSupport.stream(
+                QueryablePipeline.forTransforms(executableTransforms.keySet(), fusedComponents)
+                    .getTopologicallyOrderedTransforms()
+                    .spliterator(),
+                false)
+            .map(PTransformNode::getId)
+            .collect(Collectors.toList());
+    return Pipeline.newBuilder()
+        .setComponents(fusedComponents)
+        .addAllRootTransformIds(rootTransformIds)
+        .build();
+  }
+
   /**
-   * Return a {@link Components} like the {@code base} components, but with the only transforms
-   * equal to this fused pipeline.
+   * Return a {@link Components} like the {@code base} components, but with the set of transforms to
```

Comment:
This comment seems like it should apply toPipeline and not the private method getExecutableStages

Also, Return a {@link Pipeline}

Reply:
This comment still seems out of date since it refers to `base` components.